### PR TITLE
Make file-system based routes optional for custom servers

### PR DESCRIPTION
**Pull request by gcpantazis**
Related to #913, and making a custom routing setup more generally:

If you're using a custom server per the README and examples, you may not want to keep the file-system based routes. This commit simply puts that behind a configuration setting, so that you could turn those routes off by the following `config.next.js`:

```
module.exports = {
    useFileSystemPageRoutes: false,
}
```

If this is an acceptable solution I can expand the tests to cover it — let me know if that's worth doing. Thanks again for all the hard work!

### CODE REVIEW

File: server/index.js [hunk 1]
```diff
@@ -100,9 +100,11 @@ export default class Server {
       '/static/:path+': async (req, res, params) => {
         const p = join(this.dir, 'static', ...(params.path || []))
         await this.serveStatic(req, res, p)
-      },
+      }
+    }
 
-      '/:path*': async (req, res) => {
+    if (this.config.useFileSystemPageRoutes !== false) {
```
Comment (timneutkens) [inline server/index.js]: Think we can just remove the `!== false` and add the default here https://github.com/zeit/next.js/blob/master/server/config.js#L6

Comment (gcpantazis) [inline server/index.js]: Done!

Comment (timneutkens) [inline server/index.js]: Awesome :fire:

**Comment by rauchg** [comment]
Some people might even be interested in customizing the name of the directory, for example:

`"pagesDirectory": "screens"`

i wonder if `pagesDirectory: null` could then accomplish this? Or do you think it's too obscure? That way we have fewer settings ;)

**Comment by timneutkens** [comment]
@rauchg that's interesting. `pagesDirectory: null` would be obscure. Though `pagesDirectory: false` might just be what we want in this situation 👍

**Comment by arunoda** [comment]
I really like `#913`. But when you add a custom route it'll get the priority over the page based old routing. So, I don't think, we need a special flag to disable it.

**Comment by gcpantazis** [comment]
@rauchg I dig the direction on using this to also customize `pagesDirectory`, agreed with @timneutkens on `false` as the value. I'll hook this up and update the tests.

@arunoda Hm, perhaps I'm misunderstanding but it seems like the custom server code needs to implement `handle(req, res)` to preserve the router's plumbing, which would persist all the file-system-based routes.

**Comment by arunoda** [comment]
Since we do `handle(req, res)` at the end this is fine.
Also, it's legit for users to both types of routing at the same time.

**Comment by nkzawa** [comment]
I think the `pagesDirectory` option is not suitable for this purpose, since you might want to change the directory and want to disable file-system based routes at the same time.

As @arunoda mentioned, you don't need the flag if you define customs routes over file-system based ones, though it would be inconvenient.

Another way is to set the option as an argument of `handle` like:

```js
handle(req, res, { fileSystemRoutes: false })
```

**Comment by gcpantazis** [comment]
> I think the pagesDirectory option is not suitable for this purpose, since you might want to change the directory and want to disable file-system based routes at the same time.

Ah, of course, that makes sense @nkzawa. I'm less sure if we want to add a secondary configuration into `handle()`'s arguments when we could just centralize it into `next.config`. (The need for it at all is probably a smell that the server router API needs to be expanded to allow for removing routes directly).

For now though, I'll proceed with adding the two options into next.config:

```
{
  pagesDirectory: 'foo', // default 'pages'
  useFileSystemPublicRoutes: false // default true
}
```

(Using the key `useFileSystemPublicRoutes` just since those routes would still exist for internal plumbing / your custom server routing code, they just wouldn't be exposed to users).

**Comment by nkzawa** [comment]
Let's not include `pagesDirectory` on this PR. I think we should discuss it anyway.

**Comment by gcpantazis** [comment]
Sounds good @nkzawa, I'll split it into a separate PR and move that part of the discussion over there 👍

**Review by timneutkens — APPROVED** [review]
Looks good 👍

**Comment by timneutkens** [comment]
@nkzawa @arunoda leaving it to you to merge 👍

**Comment by timneutkens** [comment]
@arunoda can you check this one and merge if you want to take it 👍

**Comment by arunoda** [comment]
Okay. I'm good with this.

**Comment by sedubois** [comment]
Should documentation be added?

**Comment by arunoda** [comment]
@sedubois yeah! We can add it here in the options: https://github.com/zeit/next.js#custom-server-and-routing

**Pull request merged**